File: Java/BOJ/2629 양팔저울.java
import java.io.*;
import java.util.*;
import java.math.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        int[] weight = new int[N];
        for (int i = 0; i < N; i++) {
        	weight[i] = Integer.parseInt(st.nextToken());
        }
        int M = Integer.parseInt(br.readLine());
        st = new StringTokenizer(br.readLine()," ");
        int[] bead = new int[M];
        for (int i = 0 ; i < M; i++) {
        	bead[i] = Integer.parseInt(st.nextToken());
        }
        
        HashSet<Integer> dp = new HashSet<>();
        dp.add(0);
        
        for (int i : weight) {
        	HashSet<Integer> temp = new HashSet<>();
        	for(int j : dp) {
        		temp.add(i+j);
        		temp.add(Math.abs(i-j));
        	}
        	dp.addAll(temp);
        }
        List<String> result = new ArrayList<>();
        for(int i : bead) {
        	if (dp.contains(i)) {
        		result.add("Y");
        	}
        	else {
        		result.add("N");
        	}
        }
        System.out.println(String.join(" ", result));
    }
}

File: Python/BOJ/2629 양팔저울.py
import sys
N = int(sys.stdin.readline())
weight = list(map(int, sys.stdin.readline().split()))
M = list(sys.stdin.readline())
bead = list(map(int, sys.stdin.readline().split()))

dp = [0]
for i in weight:
    temp=[]
    for j in dp:
        temp += [i+j, abs(i-j)]
    dp = list(set(dp + temp))

result = []
for i in bead:
    if i in dp:
        result.append('Y')
    else:
        result.append('N')
print(*result)